File: 4week/3task/Gun/Gun.java
public class Gun {
    private int amount;
    private int capacity;
    private boolean isReady;
    private String model;
    private int totalShots;
    
    Gun(String model, int capacity) {
       this.capacity = capacity;
       this.model = model;
       this.amount = 0;
       this.totalShots = 0;
       this.isReady = false;
    }

    int getAmount() {
        return this.amount;
    }

    int getCapacity() {
        return this.capacity;
    }

    boolean ready() {
        return this.isReady;
    }
    
    String getModel() {
        return this.model;
    }

    int getTotalShots() {
        return this.totalShots;
    }

    void prepare() {
        this.isReady = true;
    }

    void reload(){
        this.amount = capacity;
    }

    void shoot() throws NotReadyException, OutOfRoundsException {
        if (this.isReady == false) {
            throw new NotReadyException ();
        }
        if (amount == 0) {
            throw new OutOfRoundsException();
        }
        System.out.println("Bang!");
        amount -=1;
        totalShots +=1;
    }
    
    public void print(){
        System.out.println("Gun model: "+this.model+"\n"
                            +"Amount: " + this.amount+"\n"
                            +"Capacity: " + this.capacity+ "\n"
                            +"Gun ready: " + this.isReady+ "\n" 
                            +"Total shots: " +this.totalShots);
    }
}

File: 4week/1task/Gun/Gun.cpp
#include <iostream>
#include "Gun.h"

Gun::Gun(const std::string& model, int capacity) {
    this->amount = capacity;;
    this->capacity = capacity;
    this->isReady = false;
    this->model = model;
    this->totalShots = 0;
}

Gun::~Gun() {}

int Gun::getAmount() const {
    return this->amount;
}

int Gun::getCapacity() const {
    return this->capacity;
}

bool Gun::ready() const {
    return this-> isReady;
}

const std::string& Gun::getModel() const {
    return this->model;
}

int Gun::getTotalShots() const {
    return this->totalShots;
}

void Gun::prepare() {
    this->isReady = true;
}

void Gun::reload() {
    this->amount = this->capacity;
    this->isReady = false;
}

void Gun::shoot() {
    if ( !ready() ) {
        throw NotReady();
    }
    if ( this->amount == 0 ) {
       throw OutOfRounds();
    }
    std::cout << "Bang!" << std::endl;
    this->amount -= 1;
    this->totalShots += 1;
}

std::ostream& operator<<(std::ostream& out, const Gun& gun) {
    out << "Gun model: " << gun.getModel() << ",\n"
    << "Amount: " << gun.getAmount() <<  ",\n"
    << "Capacity: " << gun.getCapacity() << ",\n"
    << "Gun ready: " << gun.ready() <<  ",\n"
    << "Total shots: " << gun.getTotalShots() << "." << std::endl;
    return out;
}
